### PR TITLE
Fix issues with VIA EEPROM initialization and bring more in line with rest of eeconfig functionality

### PR DESCRIPTION
**Pull request by drashna**
## Description

VIA EEPROM checks and initialization isn't touched by `eeconfig_init`.  So, it's possible that it can be ignore.  This actively happens when the EEP_RST key is hit, as well as full bootmagic's eeprom reset.

## Types of Changes

- [x] Core
- [x] Bugfix
- [x] Enhancement/optimization

## Issues Fixed or Closed by This PR

* issue reported on discord

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/via.c
Comment (fauxpark) [inline quantum/via.c:333]:
```suggestion
            uint16_t size   = dynamic_keymap_macro_get_buffer_size();
```
There's no I in ~~team~~ macro

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: quantum/via.c
Comment (drashna) [reply quantum/via.c:333]: It can't be one of my PRs unless there is a spelling error :D

**Comment by wilba** [comment]
At first glance, this changes how VIA works i.e. breaks existing functionality in `via_init()`
`via_eeprom_reset()` already calls `eeconfig_disable()`.

**Comment by drashna** [comment]
It might, let me give a second lookover.

**Comment by drashna** [comment]
This does change the functionality around a bit, namely removes `via_eeprom_reset`  However, by including the via eeprom initialization into `eeconfig_init`, then it removes the need to invalidate the via magic number.  However, the `eeconfig_init` invalidates it at the beginning, which maintains the same functionality and ensure that it will re-init it properly if power loss or the like occurred during the init process. 

It should maintain the existing behavior, but it should definitely be tested just to make sure.

File: quantum/via.c
Comment (wilba) [inline quantum/via.c]: this needs to be put back in

**Comment by wilba** [comment]
I'll review it some more later.

**Comment by wilba** [comment]
Look at #11892 which was the original code and behaviour of `id_eeprom_reset` command (i.e. it was calling `via_eeprom_reset()`.

The intent is to set the invalid state of the EEPROM prior to a reset, so that the next restart of the firmware, either the same firmware, or more importantly, new firmware, would detect the invalid state during initialization and reset EEPROM state to defaults.

Since there were some concerns with security (not going to argue about that now), the command handler was removed until such time as I can put it back in a way that pleases everyone. At the time, there didn't see much point in conditionally compiling out this feature when it wasn't even being used in VIA.

Thus, the primary purpose of the original `via_eeprom_reset()` was being called during bootmagic lite, and all it is supposed to do is set an invalid state in the EEPROM so that the next execution (esp. after flashing a new version), would detect this and reset to defaults during initialization. Part of that initialization is calling the keyboard level code to reset it's own state to defaults. This is why `via_init_kb()` is called first in `via_init()`, _before_ `via_eeprom_set_valid(true)` so that `via_eeprom_is_valid()` will still return false. FWIW VIA uses the build date as it's EEPROM magic so that firmware change can be detected. Thus even if the user doesn't use bootmagic lite with an explicit invalidating of EEPROM (i.e. a bootloader jump from physical reset button), the restarted firmware can detect a firmware change and resets EEPROM to defaults.

I don't see why this behaviour needs changing. Pulling the VIA "resetting EEPROM state" code out of `via_init()` into `eeconfig_init_via()` and calling that from the command handler only makes sense if VIA is going to get reinitialized properly at that point, including keyboard level code through `via_init_kb()`, and that handler is only expected to be called once, on startup, not in response to a command handler event.

Also, `bootmagic_lite_reset_eeprom()` was changed to not invalidate the VIA EEPROM through `via_eeprom_reset()`, i.e. `via_eeprom_set_valid(false);` doesn't get called, so bootmagic lite won't reset VIA state, since `via_eeprom_is_valid()` does not check the QMK EEPROM valid state.

It's not as easy as just calling `via_init()` from `eeconfig_init_quantum()` because `via_init()` does more than just reset EEPROM to defaults, as `via_init_kb()` can be implemented to do other startup functions and is expected to only be called once at startup.

If you can explan what the actual issues are with how it currently works, I can advise on whatever refactoring can happen in the VIA code, but right now, it's not clear.

**Comment by drashna** [comment]
> Look at #11892 which was the original code and behaviour of `id_eeprom_reset` command (i.e. it was calling `via_eeprom_reset()`.

Personally, I feel that the eeprom reset didn't need to be removed here, and is why I re-added it, but behind a define.  I can still remove it, if that's what is for the best. 

> The intent is to set the invalid state of the EEPROM prior to a reset, so that the next restart of the firmware, either the same firmware, or more importantly, new firmware, would detect the invalid state during initialization and reset EEPROM state to defaults.

Yeah, I understand that.  And that behavior is maintained.  Though, some of it is moved to re-initing the entire eeprom, not just the via specific stuff. 

> I don't see why this behaviour needs changing. Pulling the VIA "resetting EEPROM state" code out of via_init() into eeconfig_init_via() and calling that from the command handler only makes sense if VIA is going to get reinitialized properly at that point, including keyboard level code through via_init_kb(), and that handler is only expected to be called once, on startup, not in response to a command handler event.

Namely, if you hit `EEP_RST`, it does not touch the VIA config.  This is because the via eeprom stuff is not hooked into the normal eeprom initialization, and is instead handled completely separately.  Also, if full bootmagic is enabled rather than lite, the eeprom reset for that will not work.  
Which is not the correct way to handle it, IMO.   
And this isn't a theoretical issue, either.  Several people have run into this, myself included.  

instead of just adding via specific code to the keycode, it really should be handled correctly and in a way that is consistent with the rest


> It's not as easy as just calling `via_init()` from `eeconfig_init_quantum()` because `via_init()` does more than just reset EEPROM to defaults, as `via_init_kb()` can be implemented to do other startup functions and is expected to only be called once at startup.

To be blunt, `via_init_kb()` is absolutely redundant.  Its functionality can be replaced by `eeconfig_init_kb`, as well as the `matrix_init_kb`, or `keyboard_(pre|post)_init_kb` functions.  Adding a via specific init callback feels unnecessary. 


> If you can explan what the actual issues are with how it currently works, I can advise on whatever refactoring can happen in the VIA code, but right now, it's not clear.

Basically, the via eeprom initialization should be handled and hooked into the same way that every other feature in QMK that uses eeprom.   

There is no problem with having extra checks, but having it handled separately and different *has* caused issues.

**Comment by drashna** [comment]
Alternatively, I could see about just adding the dynamic keymap init stuff is added to `eeconfig_init`, but that doesn't the via stuff.

**Comment by wilba** [comment]
The code in user implementations of `via_init_kb()` could be refactored into a different `*_init_kb()` function, that hasn't been done in this PR though. The trick is to make sure that the user level init can detect an EEPROM invalid state, _especially_ the build date magic, so that the new firmware version state can be detected and handled. See `via_init_kb()` in `wt_main.c`.  As the code is now, that can't happen from `eeconfig_init_kb()` because `eeconfig_init_via()` calls `via_eeprom_set_valid(true)`.

I concede that the VIA code's usage of EEPROM is outside that of what the QMK Core code was using, it originated as being "keyboard level" customization... so `via_init_kb()` usage is being used to do things like keyboard specific initialization and when VIA is not enabled, it is called via `matrix_init_kb()` instead. It's not redundant because it's actually doing something right now. It can be refactored, though. This PR isn't fully refactoring the current behaviour that relies on `via_init_kb()` being called.

Probably the first thing to change is to make the tests of `via_eeprom_is_valid()` in places also test `eeconfig_is_enabled()`, that will then match the change made to `bootmagic_lite_reset_eeprom()`.

Probably easier if I do some refactoring and create a PR to show you what I mean.

**Comment by drashna** [comment]
Well, the via_init_kb is readded

**Comment by drashna** [comment]
`eeprom_is_enabled` and `eeprom_is_disabled` functions updated to include checks for VIA's status.

I think that pretty much covers is?

**Comment by wilba** [comment]
I'll check nothing breaks

**Comment by drashna** [comment]
Have you had a chance to verify this?

**Review by tzarc — APPROVED** [review]

**Comment by drashna** [comment]
@wilba I'd like to get this in before the final deadline for Q3 breaking changes cycle.  It should not interrupt any existing functionally, and fix issues that myself and others have run into.  

The only thing it's waiting on is feedback and/or approval from you.

**Comment by wilba** [comment]
ok will check asap

**Comment by wilba** [comment]
The current code breaks the detection of different build date firmware by `via_init_kb()` because it gets detected earlier through `matrix_init()->matrix_init_quantum()->magic()` since now `eeconfig_is_enabled()` calls `via_eeprom_is_valid()` and so `eeconfig_init()` calls `eeconfig_init_via()` before `via_init_kb()` is called.

I suggest a minor change to `keyboard.c`, moving the call to `via_init()` above `matrix_init()` so that it can detect different build date firmware.

```
void keyboard_init(void) {
    timer_init();
    sync_timer_init();
#ifdef VIA_ENABLE
    via_init();
#endif
    matrix_init();
```

I've tested this change and it works. It's the simplest fix to the problem. I will look into this problem more and refactor some of this code in the next update to VIA code, which will be after the current round of breakin changes.

**Comment by drashna** [comment]
> I suggest a minor change to `keyboard.c`, moving the call to `via_init()` above `matrix_init()` so that it can detect different build date firmware.

Applied.  And thanks!

**Pull request merged**
